scrapy/downloadermiddlewares/offsite.py

```diff
@@ -21,29 +22,34 @@
 
 
 class OffsiteMiddleware:
+    crawler: Crawler
+
+    def __init__(self, stats: StatsCollector):
+        self.stats = stats
+        self.domains_seen: set[str] = set()
+
     @classmethod
     def from_crawler(cls, crawler: Crawler) -> Self:
         assert crawler.stats
         o = cls(crawler.stats)
         crawler.signals.connect(o.spider_opened, signal=signals.spider_opened)
         crawler.signals.connect(o.request_scheduled, signal=signals.request_scheduled)
+        o.crawler = crawler
         return o
 
-    def __init__(self, stats: StatsCollector):
-        self.stats = stats
-        self.domains_seen: set[str] = set()
```

Comment:
For the record, I kind of like the `__init__` after the `from_crawler`, in order of execution :slightly_smiling_face: